File: backend/downloader.py
"""
Download manager backed by aria2c via the aria2p library.

How it works:
  1. On startup, we launch an aria2c subprocess with --enable-rpc
     (or connect to one that's already running on port 6800).
  2. Downloads are added via the JSON-RPC API using aria2p.
  3. A background polling loop runs every second, fetches all download
     statuses from aria2c, updates our in-memory cache, and fires
     optional per-download callbacks (used to sync the SQLite DB).
  4. Pause / resume / stop are async wrappers around aria2p calls that
     use asyncio.to_thread so the event loop is never blocked.

Bugs fixed vs. original:
  - pause() / resume() / stop() were synchronous but called blocking aria2p
    network I/O. When FastAPI endpoints were made async these would have
    frozen the event loop. All three are now async with asyncio.to_thread.
  - set_max_concurrent() had the same event-loop-blocking problem. Fixed.
  - shutdown() did not await the poll task cancellation, risking a race
    where the task ran one more iteration after self.api was cleared.
  - Exceptions in _sync_all / callbacks were silently swallowed with bare
    `pass`. Now logged at DEBUG so transient errors are visible.
  - stop() now also calls _clean() even when the api.get_download RPC
    fails, so internal maps never leak stale entries.
  - aria2c subprocess is killed if its RPC never becomes ready (was leaked).
"""

import asyncio
import logging
import os
import shutil
import subprocess
from time import monotonic
from typing import Callable, Dict, Optional

import aria2p

logger = logging.getLogger(__name__)

# Map aria2c status strings → our internal status vocabulary
_STATUS_MAP: dict[str, str] = {
    "active":   "downloading",
    "waiting":  "queued",
    "paused":   "paused",
    "error":    "failed",
    "complete": "completed",
    "removed":  "pending",
}


def _map_status(aria_status: str) -> str:
    return _STATUS_MAP.get(aria_status, aria_status)


class Aria2Manager:
    def __init__(self) -> None:
        self.api: Optional[aria2p.API] = None
        self._process: Optional[subprocess.Popen] = None
        self._owns_process: bool = False

        # Bidirectional mapping: our DB ids ↔ aria2 GIDs
        self._gid_map: Dict[int, str] = {}   # dl_id  → aria2 GID
        self._rev_map: Dict[str, int] = {}   # aria2 GID → dl_id

        # Per-download async callbacks fired by the poll loop → syncs DB
        self._callbacks: Dict[int, Callable] = {}

        # In-memory progress cache returned to WS & REST endpoints
        self._cache: Dict[int, dict] = {}
        self._terminal_since: Dict[int, float] = {}
        self._terminal_notified: set[int] = set()

        self._poll_task: Optional[asyncio.Task] = None
        self._poll_failures: int = 0
        self.is_running: bool = False
        self.max_concurrent: int = 3
        self.connections_per_file: int = 4
        self.terminal_cache_seconds: float = 10.0

    # ── Lifecycle ──────────────────────────────────────────────────────────────

    async def start(self, max_concurrent: int = 3, connections_per_file: int = 4) -> None:
        self.max_concurrent = max_concurrent
        self.connections_per_file = connections_per_file

        # Reuse an aria2c already running on 6800 (e.g. system daemon)
        test_api = aria2p.API(aria2p.Client(host="http://localhost", port=6800, secret=""))
        try:
            await asyncio.to_thread(test_api.get_stats)
            self.api = test_api
            self._owns_process = False
            self.is_running = True
            await self.set_max_concurrent(max_concurrent)
            self._poll_task = asyncio.create_task(self._poll_loop())
            logger.info("Reused existing aria2c daemon on port 6800.")
            return
        except Exception:
            pass

        # aria2c not running — spawn our own subprocess
        aria2c_bin = shutil.which("aria2c")
        if not aria2c_bin:
            raise RuntimeError(
                "aria2c executable not found.\n"
                "Install it first:\n"
                "  Ubuntu / Debian : sudo apt install aria2\n"
                "  macOS           : brew install aria2\n"
                "  Windows         : https://aria2.github.io/"
            )

        self._process = subprocess.Popen(
            [
                aria2c_bin,
                "--enable-rpc",
                "--rpc-listen-all=false",
                "--rpc-listen-port=6800",
                "--rpc-allow-origin-all=true",
                f"--max-concurrent-downloads={max_concurrent}",
                "--continue=true",
                "--auto-file-renaming=false",
                "--allow-overwrite=true",
                "--quiet=true",
            ],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
        self._owns_process = True

        # Poll until the RPC endpoint is ready (up to 3 s / 15 × 200 ms)
        for _ in range(15):
            await asyncio.sleep(0.2)
            try:
                new_api = aria2p.API(
                    aria2p.Client(host="http://localhost", port=6800, secret="")
                )
                await asyncio.to_thread(new_api.get_stats)
                self.api = new_api
                self._poll_failures = 0
                self.is_running = True
                self._poll_task = asyncio.create_task(self._poll_loop())
                logger.info("aria2c subprocess started; RPC ready.")
                return
            except Exception:
                continue

        # BUG FIX: kill the subprocess if we timed out — the original code leaked it.
        if self._process:
            self._process.terminate()
            self._process = None
        self._owns_process = False
        raise RuntimeError("aria2c started but its RPC server timed out after 3 s.")

    async def shutdown(self) -> None:
        self.is_running = False

        # BUG FIX: await the task cancellation so it cannot run one more tick
        # after self.api is cleared below.
        if self._poll_task:
            self._poll_task.cancel()
            try:
                await self._poll_task
            except asyncio.CancelledError:
                pass
            self._poll_task = None

        if self.api and self._owns_process:
            try:
                await asyncio.to_thread(
                    self.api.client.call, "aria2.shutdown", []
                )
            except Exception:
                pass
        self.api = None

        if self._process:
            self._process.terminate()
            self._process = None
        self._owns_process = False

    # ── Poll loop ──────────────────────────────────────────────────────────────

    async def _poll_loop(self) -> None:
        """Runs forever; fetches aria2c status every second and updates cache."""
        while self.is_running:
            await asyncio.sleep(1)
            if not self.api:
                continue
            if self._process and self._process.poll() is not None:
                await self._mark_unavailable("aria2c subprocess exited")
                return
            try:
                await self._sync_all()
                self._poll_failures = 0
            except asyncio.CancelledError:
                raise  # let cancellation propagate cleanly
            except Exception as exc:
                self._poll_failures += 1
                logger.warning(
                    "Poll loop aria2c error (%s/3): %s",
                    self._poll_failures,
                    exc,
                )
                if self._poll_failures >= 3:
                    await self._mark_unavailable("aria2c RPC stopped responding")
                    return

    async def _mark_unavailable(self, reason: str) -> None:
        logger.warning("%s; marking aria2c offline.", reason)
        self.is_running = False
        self.api = None
        self._poll_failures = 0
        if self._process and self._owns_process and self._process.poll() is None:
            self._process.terminate()
            self._process = None
            self._owns_process = False
        if self._process and self._process.poll() is not None:
            self._process = None
            self._owns_process = False
        for snap in self._cache.values():
            if snap.get("status") in {"downloading", "queued"}:
                snap["status"] = "failed"
                snap["speed"] = 0
                snap["error"] = reason
                if cb := self._callbacks.get(snap.get("id")):
                    try:
                        await cb(dict(snap))
                    except Exception as exc:
                        logger.debug("unavailable callback error for dl_id=%s: %s", snap.get("id"), exc)

    async def _sync_all(self) -> None:
        all_dls = await asyncio.to_thread(self.api.get_downloads)
        seen_gids = {dl.gid for dl in all_dls}

        for dl in all_dls:
            dl_id = self._rev_map.get(dl.gid)
            if dl_id is None:
                continue  # not one of ours

            error: Optional[str] = None
            if dl.has_failed:
                msg = getattr(dl, "error_message", None)
                code = getattr(dl, "error_code", None)
                error = (
                    str(msg) if msg
                    else (f"Error code {code}" if code else "Download failed")
                )

            cached_name = self._cache.get(dl_id, {}).get("filename", "")
            snapshot = {
                "id":               dl_id,
                "gid":              dl.gid,
                # Prefer the filename EasyDL explicitly passed to aria2c.
                # Some file hosts expose random CDN path names; aria2p may
                # report those back even when the output name is correct.
                "filename":         cached_name or dl.name,
                "status":           _map_status(dl.status),
                "bytes_downloaded": dl.completed_length,
                "total_bytes":      dl.total_length,
                "progress":         round(dl.progress, 2),
                "speed":            dl.download_speed,
                "connections":      getattr(dl, "connections", 0),
                "error":            error,
            }
            self._cache[dl_id] = snapshot

            status = snapshot["status"]
            is_terminal = status in {"completed", "failed", "pending"}
            if is_terminal:
                self._terminal_since.setdefault(dl_id, monotonic())
            else:
                self._terminal_since.pop(dl_id, None)
                self._terminal_notified.discard(dl_id)

            should_notify = not is_terminal or dl_id not in self._terminal_notified
            if should_notify and (cb := self._callbacks.get(dl_id)):
                try:
                    await cb(snapshot)
                    if is_terminal:
                        self._terminal_notified.add(dl_id)
                except Exception as exc:
                    logger.debug("on_update callback error for dl_id=%s: %s", dl_id, exc)

            if is_terminal:
                seen_at = self._terminal_since.get(dl_id, monotonic())
                if monotonic() - seen_at >= self.terminal_cache_seconds:
                    self._clean(dl_id, dl.gid)

        for dl_id, gid in list(self._gid_map.items()):
            if gid not in seen_gids:
                snap = {
                    "id": dl_id,
                    "gid": gid,
                    "filename": self._cache.get(dl_id, {}).get("filename", ""),
                    "status": "failed",
                    "bytes_downloaded": self._cache.get(dl_id, {}).get("bytes_downloaded", 0),
                    "total_bytes": self._cache.get(dl_id, {}).get("total_bytes", 0),
                    "progress": self._cache.get(dl_id, {}).get("progress", 0),
                    "speed": 0,
                    "connections": 0,
                    "error": "aria2c no longer tracks this download",
                }
                if cb := self._callbacks.get(dl_id):
                    try:
                        await cb(snap)
                    except Exception as exc:
                        logger.debug("missing-gid callback error for dl_id=%s: %s", dl_id, exc)
                self._clean(dl_id, gid)

    # ── Enqueue ────────────────────────────────────────────────────────────────

    async def enqueue(
        self,
        dl_id: int,
        url: str,
        output_path: str,
        on_update: Optional[Callable] = None,
        fresh_start: bool = False,
    ) -> str:
        """
        Add a URL to aria2c's queue. Returns the aria2 GID immediately.
        aria2c handles the download; we watch via the poll loop.
        """
        if not self.api:
            raise RuntimeError("aria2c is not running. Check server logs.")
        if dl_id in self._gid_map:
            raise RuntimeError("Download is already queued in aria2c.")

        out_dir = os.path.dirname(output_path) or "."
        os.makedirs(out_dir, exist_ok=True)

        if fresh_start:
            await asyncio.to_thread(self._discard_partial_files, output_path)

        # FuckingFast's file backend does not reliably honor bounded range
        # requests such as bytes=100-200; it often returns bytes=100-EOF.
        # aria2's segmented mode treats that as a corrupt response, so use one
        # connection per file for stable downloads.
        connections = 1
        options = {
            "dir":                       out_dir,
            "out":                       os.path.basename(output_path),
            "continue":                  "false" if fresh_start else "true",
            "split":                     str(connections),
            "min-split-size":            "1M",
            "max-connection-per-server": str(connections),
        }

        dl = await asyncio.to_thread(self.api.add_uris, [url], options=options)
        gid = dl.gid

        self._gid_map[dl_id] = gid
        self._rev_map[gid] = dl_id
        self._terminal_since.pop(dl_id, None)
        self._terminal_notified.discard(dl_id)
        if on_update:
            self._callbacks[dl_id] = on_update

        # Pre-populate cache so the UI shows "queued" immediately
        self._cache[dl_id] = {
            "id":               dl_id,
            "gid":              gid,
            "filename":         os.path.basename(output_path),
            "status":           "queued",
            "bytes_downloaded": 0,
            "total_bytes":      0,
            "progress":         0.0,
            "speed":            0,
            "connections":      0,
            "error":            None,
        }
        return gid

    @staticmethod
    def _discard_partial_files(output_path: str) -> None:
        for path in (output_path, f"{output_path}.aria2"):
            try:
                if os.path.isfile(path):
                    os.remove(path)
            except OSError as exc:
                logger.warning("Could not remove partial download file %s: %s", path, exc)

    # ── Controls ───────────────────────────────────────────────────────────────
    # BUG FIX: all three were synchronous and called blocking aria2p network I/O.
    # When invoked from async FastAPI endpoints they would have frozen the event
    # loop. Converted to async with asyncio.to_thread for every RPC call.

    async def pause(self, dl_id: int) -> bool:
        if not self.api:
            raise RuntimeError("aria2c is not running")
        gid = self._gid_map.get(dl_id)
        if not gid:
            return False
        try:
            dl_obj = await asyncio.to_thread(self.api.get_download, gid)
            await asyncio.to_thread(self.api.pause, [dl_obj])
            if dl_id in self._cache:
                self._cache[dl_id]["status"] = "paused"
                self._cache[dl_id]["speed"] = 0
            return True
        except Exception as exc:
            logger.warning("pause(%s) failed: %s", dl_id, exc)
            raise

    async def resume(self, dl_id: int) -> None:
        if not self.api:
            return
        gid = self._gid_map.get(dl_id)
        if not gid:
            return
        try:
            dl_obj = await asyncio.to_thread(self.api.get_download, gid)
            await asyncio.to_thread(self.api.resume, [dl_obj])
            if dl_id in self._cache:
                self._cache[dl_id]["status"] = "downloading"
        except Exception as exc:
            logger.warning("resume(%s) failed: %s", dl_id, exc)
            raise  # Re-raise so API handlers can catch and handle

    async def stop(self, dl_id: int) -> None:
        gid = self._gid_map.get(dl_id)
        if self.api and gid:
            try:
                dl_obj = await asyncio.to_thread(self.api.get_download, gid)
                await asyncio.to_thread(self.api.remove, [dl_obj])
            except Exception as exc:
                # BUG FIX: original code returned without calling _clean on failure,
                # leaving stale entries in _gid_map / _rev_map / _cache forever.
                logger.warning("stop(%s) aria2c remove failed: %s", dl_id, exc)
        self._clean(dl_id, gid)  # always clean regardless of RPC outcome

    def _clean(self, dl_id: int, gid: Optional[str]) -> None:
        self._gid_map.pop(dl_id, None)
        if gid:
            self._rev_map.pop(gid, None)
        self._callbacks.pop(dl_id, None)
        self._cache.pop(dl_id, None)
        self._terminal_since.pop(dl_id, None)
        self._terminal_notified.discard(dl_id)

    # ── Config ─────────────────────────────────────────────────────────────────

    async def set_max_concurrent(self, n: int) -> None:
        """BUG FIX: was sync; calling blocking client.call() from async context
        would freeze the event loop. Now uses asyncio.to_thread."""
        self.max_concurrent = max(1, n)
        if self.api:
            try:
                await asyncio.to_thread(
                    self.api.client.call,
                    "aria2.changeGlobalOption",
                    [{"max-concurrent-downloads": str(n)}],
                )
            except Exception as exc:
                logger.warning("set_max_concurrent RPC failed: %s", exc)

    def set_connections_per_file(self, n: int) -> None:
        # No RPC call needed — applied to the next enqueue() call.
        self.connections_per_file = max(1, min(16, n))

    # ── Queries ────────────────────────────────────────────────────────────────

    def get_progress(self, dl_id: int) -> Optional[dict]:
        return self._cache.get(dl_id)

    def all_progress(self) -> list:
        return list(self._cache.values())

    async def global_stats(self) -> dict:
        if not self.api:
            return {"aria2_running": False}
        try:
            s = await asyncio.to_thread(self.api.get_stats)
            return {
                "aria2_running":  True,
                "download_speed": s.download_speed,
                "upload_speed":   s.upload_speed,
                "num_active":     s.num_active,
                "num_waiting":    s.num_waiting,
                "num_stopped":    s.num_stopped_total,
            }
        except Exception as exc:
            logger.warning("aria2c stats unavailable: %s", exc)
            return {"aria2_running": False}


# Module-level singleton used throughout the app
dm = Aria2Manager()

File: src/downloader.ts
import { spawn, type ChildProcess } from 'child_process';
import { execFile as execFileCallback } from 'child_process';
import fs from 'fs';
import path from 'path';
import { promisify } from 'util';

const execFile = promisify(execFileCallback);
const RPC_URL = 'http://localhost:6800/jsonrpc';

type ProgressCallback = (snap: ProgressSnapshot) => Promise<void>;

export interface ProgressSnapshot {
  id: number;
  gid: string;
  filename: string;
  status: string;
  bytes_downloaded: number;
  total_bytes: number;
  progress: number;
  speed: number;
  connections: number;
  error: string | null;
}

const STATUS_MAP: Record<string, string> = {
  active: 'downloading',
  waiting: 'queued',
  paused: 'paused',
  error: 'failed',
  complete: 'completed',
  removed: 'pending',
};

let rpcIdCounter = 1;

async function rpc(method: string, params: unknown[] = []): Promise<unknown> {
  const id = String(rpcIdCounter++);
  const resp = await fetch(RPC_URL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ jsonrpc: '2.0', id, method, params }),
  });
  if (!resp.ok) throw new Error(`RPC HTTP ${resp.status}`);
  const data = await resp.json() as { result?: unknown; error?: { message: string } };
  if (data.error) throw new Error(data.error.message);
  return data.result;
}

function parseAriaInt(value: string | undefined): number {
  return Number.parseInt(value ?? '0', 10) || 0;
}

async function findAria2c(): Promise<string> {
  try {
    const { stdout } = await execFile(process.platform === 'win32' ? 'where' : 'which', ['aria2c']);
    const first = stdout.split(/\r?\n/).map(line => line.trim()).filter(Boolean)[0];
    if (first) return first;
  } catch {
    // Fall through to the helpful error below.
  }
  throw new Error(
    'aria2c executable not found.\n' +
    'Install it first:\n' +
    '  Ubuntu/Debian: sudo apt install aria2\n' +
    '  macOS:         brew install aria2\n' +
    '  Windows:       https://aria2.github.io/',
  );
}

class Aria2Manager {
  private process: ChildProcess | null = null;
  private ownsProcess = false;
  private gidMap = new Map<number, string>();
  private revMap = new Map<string, number>();
  private callbacks = new Map<number, ProgressCallback>();
  private cache = new Map<number, ProgressSnapshot>();
  private terminalSince = new Map<number, number>();
  private terminalNotified = new Set<number>();
  private pollTimer: ReturnType<typeof setInterval> | null = null;
  private pollRunning = false;
  private pollFailures = 0;

  isRunning = false;
  maxConcurrent = 3;
  connectionsPerFile = 4;
  readonly terminalCacheMs = 10000;

  async start(maxConcurrent = 3, connectionsPerFile = 4): Promise<void> {
    this.maxConcurrent = maxConcurrent;
    this.connectionsPerFile = connectionsPerFile;

    try {
      await rpc('aria2.getGlobalStat');
      this.isRunning = true;
      this.ownsProcess = false;
      await this.setMaxConcurrent(maxConcurrent);
      this.startPollTimer();
      console.log('Reused existing aria2c daemon on port 6800.');
      return;
    } catch {
      // Start our own daemon below.
    }

    const aria2cBin = await findAria2c();
    this.process = spawn(aria2cBin, [
      '--enable-rpc',
      '--rpc-listen-all=false',
      '--rpc-listen-port=6800',
      '--rpc-allow-origin-all=true',
      `--max-concurrent-downloads=${maxConcurrent}`,
      '--continue=true',
      '--auto-file-renaming=false',
      '--allow-overwrite=true',
      '--quiet=true',
    ], { stdio: 'ignore', windowsHide: true });
    this.ownsProcess = true;

    for (let i = 0; i < 15; i++) {
      await new Promise(resolve => setTimeout(resolve, 200));
      try {
        await rpc('aria2.getGlobalStat');
        this.isRunning = true;
        this.pollFailures = 0;
        this.startPollTimer();
        console.log('aria2c subprocess started; RPC ready.');
        return;
      } catch {
        continue;
      }
    }

    this.process.kill();
    this.process = null;
    this.ownsProcess = false;
    throw new Error('aria2c started but its RPC server timed out after 3 s.');
  }

  async shutdown(): Promise<void> {
    this.isRunning = false;
    if (this.pollTimer) {
      clearInterval(this.pollTimer);
      this.pollTimer = null;
    }
    if (this.ownsProcess) {
      try {
        await rpc('aria2.shutdown');
      } catch {
        // Ignore shutdown races.
      }
    }
    this.process?.kill();
    this.process = null;
    this.ownsProcess = false;
  }

  private startPollTimer(): void {
    if (this.pollTimer) clearInterval(this.pollTimer);
    this.pollTimer = setInterval(() => {
      this.pollLoop().catch(err => {
        console.warn('Poll loop failed:', err);
      });
    }, 1000);
  }

  private async pollLoop(): Promise<void> {
    if (!this.isRunning || this.pollRunning) return;
    if (this.process && this.ownsProcess && this.process.exitCode !== null) {
      await this.markUnavailable('aria2c subprocess exited');
      return;
    }
    this.pollRunning = true;
    try {
      await this.syncAll();
      this.pollFailures = 0;
    } catch (err) {
      this.pollFailures++;
      console.warn(`Poll loop aria2c error (${this.pollFailures}/3):`, err);
      if (this.pollFailures >= 3) await this.markUnavailable('aria2c RPC stopped responding');
    } finally {
      this.pollRunning = false;
    }
  }

  private async markUnavailable(reason: string): Promise<void> {
    console.warn(`${reason}; marking aria2c offline.`);
    this.isRunning = false;
    if (this.pollTimer) {
      clearInterval(this.pollTimer);
      this.pollTimer = null;
    }
    this.process?.kill();
    this.process = null;
    this.ownsProcess = false;
    this.pollFailures = 0;

    for (const snap of this.cache.values()) {
      if (['downloading', 'queued'].includes(snap.status)) {
        snap.status = 'failed';
        snap.speed = 0;
        snap.error = reason;
        const cb = this.callbacks.get(snap.id);
        if (cb) {
          try {
            await cb({ ...snap });
          } catch (err) {
            console.debug(`unavailable callback error for dl_id=${snap.id}:`, err);
          }
        }
      }
    }
  }

  private async syncAll(): Promise<void> {
    type AriaDownload = {
      gid: string;
      status: string;
      completedLength: string;
      totalLength: string;
      downloadSpeed: string;
      connections?: string;
      errorMessage?: string;
      errorCode?: string;
      files?: Array<{ path: string }>;
    };
    const keys = ['gid', 'status', 'completedLength', 'totalLength', 'downloadSpeed', 'connections', 'errorMessage', 'errorCode', 'files'];
    const active = await rpc('aria2.tellActive', [keys]) as AriaDownload[];
    const waiting = await rpc('aria2.tellWaiting', [0, 1000, keys]) as AriaDownload[];
    const stopped = await rpc('aria2.tellStopped', [0, 1000, keys]) as AriaDownload[];
    const all = [...active, ...waiting, ...stopped];
    const seenGids = new Set(all.map(dl => dl.gid));

    for (const dl of all) {
      const dlId = this.revMap.get(dl.gid);
      if (dlId === undefined) continue;

      const status = STATUS_MAP[dl.status] ?? dl.status;
      const error = dl.errorMessage
        ? dl.errorMessage
        : dl.errorCode
          ? `Error code ${dl.errorCode}`
          : status === 'failed'
            ? 'Download failed'
            : null;
      const cachedFilename = this.cache.get(dlId)?.filename ?? '';
      const ariaFilename = dl.files?.[0]?.path ? path.basename(dl.files[0].path) : '';
      const bytesDownloaded = parseAriaInt(dl.completedLength);
      const totalBytes = parseAriaInt(dl.totalLength);
      const snap: ProgressSnapshot = {
        id: dlId,
        gid: dl.gid,
        filename: cachedFilename || ariaFilename,
        status,
        bytes_downloaded: bytesDownloaded,
        total_bytes: totalBytes,
        progress: totalBytes > 0 ? Math.round((bytesDownloaded / totalBytes) * 10000) / 100 : 0,
        speed: parseAriaInt(dl.downloadSpeed),
        connections: parseAriaInt(dl.connections),
        error,
      };
      this.cache.set(dlId, snap);

      const isTerminal = ['completed', 'failed', 'pending'].includes(status);
      if (isTerminal) {
        if (!this.terminalSince.has(dlId)) this.terminalSince.set(dlId, Date.now());
      } else {
        this.terminalSince.delete(dlId);
        this.terminalNotified.delete(dlId);
      }

      const shouldNotify = !isTerminal || !this.terminalNotified.has(dlId);
      if (shouldNotify) {
        const cb = this.callbacks.get(dlId);
        if (cb) {
          try {
            await cb({ ...snap });
            if (isTerminal) this.terminalNotified.add(dlId);
          } catch (err) {
            console.debug(`on_update callback error for dl_id=${dlId}:`, err);
          }
        }
      }

      if (isTerminal) {
        const since = this.terminalSince.get(dlId) ?? Date.now();
        if (Date.now() - since >= this.terminalCacheMs) this.clean(dlId, dl.gid);
      }
    }

    for (const [dlId, gid] of this.gidMap) {
      if (seenGids.has(gid)) continue;
      const old = this.cache.get(dlId);
      const snap: ProgressSnapshot = {
        id: dlId,
        gid,
        filename: old?.filename ?? '',
        status: 'failed',
        bytes_downloaded: old?.bytes_downloaded ?? 0,
        total_bytes: old?.total_bytes ?? 0,
        progress: old?.progress ?? 0,
        speed: 0,
        connections: 0,
        error: 'aria2c no longer tracks this download',
      };
      const cb = this.callbacks.get(dlId);
      if (cb) {
        try {
          await cb(snap);
        } catch (err) {
          console.debug(`missing-gid callback error for dl_id=${dlId}:`, err);
        }
      }
      this.clean(dlId, gid);
    }
  }

  async enqueue(
    dlId: number,
    url: string,
    outputPath: string,
    onUpdate?: ProgressCallback,
    freshStart = false,
  ): Promise<string> {
    if (!this.isRunning) throw new Error('aria2c is not running. Check server logs.');
    if (this.gidMap.has(dlId)) throw new Error('Download is already queued in aria2c.');

    fs.mkdirSync(path.dirname(outputPath), { recursive: true });
    if (freshStart) this.discardPartialFiles(outputPath);

    // FuckingFast frequently mishandles bounded range requests, so a single
    // connection is more reliable than segmented downloads for this host.
    const connections = 1;
    const gid = await rpc('aria2.addUri', [
      [url],
      {
        dir: path.dirname(outputPath),
        out: path.basename(outputPath),
        continue: freshStart ? 'false' : 'true',
        split: String(connections),
        'min-split-size': '1M',
        'max-connection-per-server': String(connections),
      },
    ]) as string;

    this.gidMap.set(dlId, gid);
    this.revMap.set(gid, dlId);
    this.terminalSince.delete(dlId);
    this.terminalNotified.delete(dlId);
    if (onUpdate) this.callbacks.set(dlId, onUpdate);

    this.cache.set(dlId, {
      id: dlId,
      gid,
      filename: path.basename(outputPath),
      status: 'queued',
      bytes_downloaded: 0,
      total_bytes: 0,
      progress: 0,
      speed: 0,
      connections: 0,
      error: null,
    });
    return gid;
  }

  private discardPartialFiles(outputPath: string): void {
    for (const filePath of [outputPath, `${outputPath}.aria2`]) {
      try {
        if (fs.existsSync(filePath) && fs.statSync(filePath).isFile()) fs.rmSync(filePath);
      } catch (err) {
        console.warn(`Could not remove partial download file ${filePath}:`, err);
      }
    }
  }

  async pause(dlId: number): Promise<boolean> {
    if (!this.isRunning) throw new Error('aria2c is not running');
    const gid = this.gidMap.get(dlId);
    if (!gid) return false;
    try {
      await rpc('aria2.pause', [gid]);
      const snap = this.cache.get(dlId);
      if (snap) {
        snap.status = 'paused';
        snap.speed = 0;
      }
      return true;
    } catch (err) {
      console.warn(`pause(${dlId}) failed:`, err);
      throw err;
    }
  }

  async resume(dlId: number): Promise<void> {
    if (!this.isRunning) return;
    const gid = this.gidMap.get(dlId);
    if (!gid) return;
    await rpc('aria2.unpause', [gid]);
    const snap = this.cache.get(dlId);
    if (snap) snap.status = 'downloading';
  }

  async stop(dlId: number): Promise<void> {
    const gid = this.gidMap.get(dlId);
    if (gid && this.isRunning) {
      try {
        await rpc('aria2.remove', [gid]);
      } catch (err) {
        console.warn(`stop(${dlId}) aria2c remove failed:`, err);
      }
    }
    this.clean(dlId, gid ?? null);
  }

  private clean(dlId: number, gid: string | null): void {
    this.gidMap.delete(dlId);
    if (gid) this.revMap.delete(gid);
    this.callbacks.delete(dlId);
    this.cache.delete(dlId);
    this.terminalSince.delete(dlId);
    this.terminalNotified.delete(dlId);
  }

  async setMaxConcurrent(n: number): Promise<void> {
    this.maxConcurrent = Math.max(1, n);
    if (!this.isRunning) return;
    try {
      await rpc('aria2.changeGlobalOption', [{ 'max-concurrent-downloads': String(this.maxConcurrent) }]);
    } catch (err) {
      console.warn('setMaxConcurrent RPC failed:', err);
    }
  }

  setConnectionsPerFile(n: number): void {
    this.connectionsPerFile = Math.max(1, Math.min(16, n));
  }

  getProgress(dlId: number): ProgressSnapshot | undefined {
    return this.cache.get(dlId);
  }

  allProgress(): ProgressSnapshot[] {
    return [...this.cache.values()].map(snap => ({ ...snap }));
  }

  async globalStats(): Promise<Record<string, unknown>> {
    if (!this.isRunning) return { aria2_running: false };
    try {
      const s = await rpc('aria2.getGlobalStat') as {
        downloadSpeed: string;
        uploadSpeed: string;
        numActive: string;
        numWaiting: string;
        numStoppedTotal: string;
      };
      return {
        aria2_running: true,
        download_speed: parseAriaInt(s.downloadSpeed),
        upload_speed: parseAriaInt(s.uploadSpeed),
        num_active: parseAriaInt(s.numActive),
        num_waiting: parseAriaInt(s.numWaiting),
        num_stopped: parseAriaInt(s.numStoppedTotal),
      };
    } catch (err) {
      console.warn('aria2c stats unavailable:', err);
      return { aria2_running: false };
    }
  }
}

export const dm = new Aria2Manager();
